book/html.md

```diff
@@ -5,408 +5,559 @@ prev: text
+of characters into an HTML tree, no matter how confusing the markup.[^3]
+The full algorithm is, as you might expect, complicated beyond belief,
+with dozens of ever-more-special cases forming a taxonomy of human
+error, but one of the nicer time-saving innovations is *implicit* tags.
+
+[^3]: Yes, it's crazy, and for a few years in the early '00s the W3C
+    tried to [do away with it](https://www.w3.org/TR/xhtml1/). They
+    failed.
+
+Normally, an HTML document starts with a familiar boilerplate:
+
+``` {.html}
+<!doctype html>
+<html>
+  <head>
+  </head>
+  <body>
+  </body>
+</html>
 ```
 
-One complication is `layout`\'s local variables: the current `x` and `y`
-position, the current `bold` and `italic` state, and even that tricky
-`terminal_space` variable that took up so much time in the last chapter.
-Plus, there\'s the `display_list` variable for render commands. To turn
-`layout` into a recursive function, we\'ll need those local variables to
-become global state that multiple invocations of `layout` can share:[^6]
+In reality, *all six* of these tags, except the doctype, are optional:
+browsers insert them automatically. To do so, they compare the current
+token's tag to the list of currently open elements; that reveals
+whether any additional elements need to be created:
 
-``` {.python}
-class State:
-    def __init__(self):
-        self.x = 13
-        self.y = 13
-        self.bold = False
-        self.italic = False
-        self.terminal_space = True
-        self.dl = []
+``` {.python indent=4}
+for tok in tokens:
+    implicit_tags(tok, currently_open)
+    # ...
+```
+
+The `implicit_tags` function examines the incoming token and
+determines whether any implicit tags need to be added. Importantly,
+more than one implicit tag may need to be added, so the function uses
+a loop:
 
-state = State()
+``` {.python}
+def implicit_tags(tok, currently_open):
+    tag = tok.tag if isinstance(tok, Tag) else None
+    while True:
+        open_tags = [node.tag for node in currently_open]
+        # ...
 ```
 
-The local variable `x` is now replaced by the field `state.x`.
+That loop needs to handle each possible implicit tag. The implicit
+`<html>` tag is easiest, since it must be the root element:
 
-The the pieces of the old `layout` function must now migrate to
-`layout_open`, `layout_close`, and `layout_text`. For example,
-`layout_open` needs to set the bold and italics flags.
+``` {.python indent=8}
+if open_tags == [] and tag != "html":
+    currently_open.append(ElementNode("html"))
+```
+
+With the `<head>` and `<body>` elements, you need to look at the tag
+being handled. Some tags go in the `<head>` element:
 
 ``` {.python}
-def layout_open(node):
-    if node.tag == "b":
-        state.bold = True
-    elif node.tag == "i":
-        state.italic = True
+HEAD_TAGS = [
+    "base", "basefont", "bgsound", "noscript",
+    "link", "meta", "title", "style", "script",
```

Comment:
script can go in body; I guess it's only when these tags are encountered outside of body/head that this behavior happens?

Reply:
Yeah, but if you encounter script before either head or body is opened, HTML5 parsing will create a head node. HTML5 more or less enforces the requirement that every DOM tree starts with HTML, HEAD, and BODY, and that HEAD only contains one of those elements (except under noscript), but as you say BODY can also contain script and noscript.